website/blog/2023-05-01-evolving-data-engineer-craft.md

```diff
@@ -0,0 +1,200 @@
+          - type: check       # not supported -- will warn & skip
+            expression: "id > 0"
+        tests:
+          - unique            # primary_key constraint is not enforced
+      - name: customer_name
+        data_type: text
+      - name: first_transaction_date
+        data_type: date
+```
+
+```sql
+--Exact SQL run against database
+create or replace transient table <database>.<schema>.dim_customers        
+(
+    id integer not null primary key,
+    customer_name text,
+    first_transaction_date date  
+)
+as
+(
+select 
+  1 as id, 
+  'My Favorite Customer' as customer_name, 
+  cast('2019-01-01' as date) as first_transaction_date
+);
+```
+
+[**Model Versions**](https://docs.getdbt.com/docs/collaborate/publish/model-versions)
+
+- **Problems you’re living through**: I change my vital model `fct_orders.sql` every week, and many people rely on this for their work. However, I keep getting doubtful questions on what’s changed since my last update, and I don’t have great way to instill confidence that this will/won’t break the way they rely on it.
+- **What it is**: Model versions in v1.5 allow you to create, specify, and reference versions of models. Core reporting models can now be updated and deprecated following software engineering practices and create systems of accountability between data creators and data consumers.
+- **How it will change your daily work**: Not every model is going to need to be versioned, but for core models that power your business intelligence, *power your data team*, you will now have the option to create multiple versions of a model and implement breaking changes in a more realistic and accountable way. Say I’m the primary owner of Core Data Team dbt `Project A`, and inside that project contains a core `dim_customers` model that powers the way finance, customer success, and marketing analyze customer data and CLV(customer lifetime value). I need to make a breaking change to `dim_customers`—CLV is going to be removed in favor of a more complex ROI value. Finance team uses the existing CLV value for cohort analysis and other reports, but understands that the new ROI column may be more favorable over time. However, it takes time to transition those reports and systems to conform to the ROI values, so `Project A` can develop a `dim_customers__v2` that drops LTV in favor for the new ROI.
+
+```yaml
+models:
+  - name: dim_customers
+    latest_version: 2
+    config:
+      materialized: table
+      test: test
+    tests:
+      - unique:
+          column_name: "location_id"
+    columns:
+      - name: customer_id
+        description: This is the primary key
+        data_type: int
+      - name: location_id
+        description: id of location
+        data_type: int
+        tests: 
+          - not_null
+    versions:
+      - v: 3
+        columns: 
+          - include: '*'
+            exclude: ['location_id']
+          - name: extra
```

Comment:
I'd recommend adding the `data_type` here. Otherwise, when a contract is enabled on this model, there will be an error message (and it won't be obvious to the user what to do to fix it).

```suggestion
          - name: extra
            data_type: text
```